docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,232 @@
+- Start Date: 2021-02-08
+- FLIP PR: [#1011](https://github.com/google/flax/pull/1011)
+- FLIP Issue: [#1009](https://github.com/google/flax/issues/1009)
+
+Table of contents:
+
+- [Summary]
+- [Functional API]
+- [Previous API]
+- [Linen Helper]
+
+# Summary
+[summary]: #summary
+
+Our current API (referred to as [previous API] in this document) uses a pattern
+where an `Optimizer` dataclass is created from a pytree of `target` variables
+and from an `OptimizerDef` that defines how to update optimizer state,
+hyperparameters, and target variables. This pattern is relatively complex for
+implementing a simple optimizer, while being quite verbose in the typical Linen
+train step (especially when using mutable state collections).
+
+This FLIP proposes a new purely [functional API] where optimizers are simply
+defined with a `init()` and an `update()` function and a new [Linen helper] that
+reduces boilerplate in a typical Linen train step.
+
+
+# Functional API
```

Comment:
Could you add a subsection comparing this to Optax API?

Reply:
Yes, I would like to understand what a version where we adopt Optax looks like -- if you directly use the Optax API, how would this proposal change? Would we still need some utilities or a wrapper due to how we represent variable collections?

Reply:
Refactored the entire FLIP. It now starts with Optax in the motivation.

Reply:
Optax can be used with a tiny wrapper.